Dockerfiles/agent/windows/entrypoint/Service.cpp

```diff
@@ -0,0 +1,155 @@
+{
+    CloseServiceHandle(_scManagerHandle);
+    CloseServiceHandle(_serviceHandle);
+}
+
+DWORD Service::PID()
+{
+    return _processId;
+}
+
+void Service::Start(std::chrono::milliseconds timeout)
+{
+    if (!StartService(_serviceHandle, 0, nullptr))
+    {
+        const DWORD lastError = GetLastError();
+        if (lastError != ERROR_SERVICE_ALREADY_RUNNING)
+        {
+            throw Win32Exception("Could not start the service");
+        }
+    }
+
+    SERVICE_STATUS_PROCESS serviceStatus;
+    do
+    {
+        DWORD unused = 0;
+        if (!QueryServiceStatusEx(_serviceHandle,
+            SC_STATUS_PROCESS_INFO,
+            reinterpret_cast<LPBYTE>(&serviceStatus),
+            sizeof(SERVICE_STATUS_PROCESS),
+            &unused))
+        {
+            throw Win32Exception("Could not query the service status");
+        }
+        
+        if (serviceStatus.dwCurrentState != SERVICE_RUNNING)
+        {
+            timeout -= std::chrono::seconds(1);
+            Sleep(1000);
+            if (timeout.count() <= 0)
+            {
+                throw std::exception("Timeout while starting the service");
+            }
+        }
+    } while (serviceStatus.dwCurrentState != SERVICE_RUNNING);
+    _processId = serviceStatus.dwProcessId;
+}
+
+void Service::Stop(std::chrono::milliseconds timeout)
+{
+    DWORD sizeNeededDependentServices;
+    DWORD countDependentServices;
+
+    if (!EnumDependentServices(
+        _serviceHandle,
+        SERVICE_ACTIVE,
+        nullptr,
+        0,
+        &sizeNeededDependentServices,
+        &countDependentServices))
+    {
+        // If the Enum call fails, then there are dependent services to be stopped first
+        if (GetLastError() != ERROR_MORE_DATA)
+        {
+            // The last error must be ERROR_MORE_DATA
+            throw Win32Exception("Unexpected error while fetching dependent services");
+        }
+
+        ENUM_SERVICE_STATUS_PTR depSvcs(
+            static_cast<LPENUM_SERVICE_STATUS>(
+                HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeNeededDependentServices)), heapFree);
+
+        if (!EnumDependentServices(
+            _serviceHandle,
+            SERVICE_ACTIVE,
+            depSvcs.get(),
+            sizeNeededDependentServices,
+            &sizeNeededDependentServices,
+            &countDependentServices))
+        {
+            throw Win32Exception("Could not enumerate dependent services");
+        }
+        for (DWORD i = 0; i < countDependentServices; ++i)
+        {
+            // Note that by giving dependent services the same timeout
+            // we may exceed our timeout ourselves.
+            Service(depSvcs.get()[i].lpServiceName).Stop(timeout);
+        }
+    }
+
+    SERVICE_STATUS_PROCESS serviceStatus;
+    if (!ControlService(_serviceHandle, SERVICE_CONTROL_STOP, reinterpret_cast<LPSERVICE_STATUS>(&serviceStatus)))
+    {
+        throw Win32Exception("Could not stop the service");
+    }
```

Comment:
nit: This is not a problem now but for some dependency graphs you might try to stop a service twice and this would fail with `ERROR_SERVICE_CANNOT_ACCEPT_CTRL`. Not sure if the extra complexity for handling that unlikely situation is worth it though.

Reply:
You don't need cycles: if you have three services `A`, `B` and `C` such that: 
- `B` depends on `A` and 
- `C` depends on `B` and `A`,

and you try to stop `A`, you would stop service `C` twice (one because the direct dependency and the other one because of `B`'s dependency).